File: examples/inefficient_example.js
/**
 * Example of inefficient JavaScript code patterns.
 * These patterns should be avoided for better performance.
 */

/**
 * INEFFICIENT: Modifying DOM in a loop
 * Causes multiple reflows and repaints
 */
function inefficientDOMManipulation(items) {
    const container = document.getElementById('container');
    for (let i = 0; i < items.length; i++) {
        const div = document.createElement('div');
        div.textContent = items[i];
        container.appendChild(div);  // Multiple DOM modifications
    }
}

/**
 * INEFFICIENT: Not caching array length in loop
 */
function inefficientArrayIteration(arr) {
    let sum = 0;
    for (let i = 0; i < arr.length; i++) {  // arr.length evaluated each iteration
        sum += arr[i];
    }
    return sum;
}

/**
 * INEFFICIENT: Creating functions inside loops
 */
function inefficientFunctionCreation(items) {
    const handlers = [];
    for (let i = 0; i < items.length; i++) {
        handlers.push(function() {  // New function object created each iteration
            console.log(items[i]);
        });
    }
    return handlers;
}

/**
 * INEFFICIENT: Using delete operator on array elements
 * Creates holes in array and prevents optimization
 */
function inefficientArrayDeletion(arr, index) {
    delete arr[index];  // Creates sparse array
    return arr;
}

/**
 * INEFFICIENT: Multiple array iterations
 */
function inefficientMultipleIterations(data) {
    const filtered = data.filter(x => x > 0);
    const squared = filtered.map(x => x * x);
    const sum = squared.reduce((a, b) => a + b, 0);
    return sum;
}

/**
 * INEFFICIENT: String concatenation in loop
 */
function inefficientStringConcat(items) {
    let result = '';
    for (let i = 0; i < items.length; i++) {
        result += items[i] + ',';  // Creates new string each time
    }
    return result;
}

/**
 * INEFFICIENT: Using Array.indexOf for membership testing
 */
function inefficientMembershipTest(arr, searchItems) {
    const found = [];
    for (let item of searchItems) {
        if (arr.indexOf(item) !== -1) {  // O(n) search for each item
            found.push(item);
        }
    }
    return found;
}

/**
 * INEFFICIENT: Nested loops without optimization
 */
function inefficientNestedLoops(arr1, arr2) {
    const matches = [];
    for (let i = 0; i < arr1.length; i++) {
        for (let j = 0; j < arr2.length; j++) {
            if (arr1[i] === arr2[j]) {
                matches.push(arr1[i]);
            }
        }
    }
    return matches;
}

/**
 * INEFFICIENT: Not using const/let properly
 */
var globalVar = 0;  // Global variable
function inefficientVariableScope() {
    for (var i = 0; i < 100; i++) {  // var leaks to function scope
        globalVar += i;
    }
    return globalVar;
}

/**
 * INEFFICIENT: Synchronous operations that could be async
 */
function inefficientSyncOperation(data) {
    // Simulating blocking operation
    const start = Date.now();
    while (Date.now() - start < 1000) {
        // Blocking the event loop
    }
    return data;
}

/**
 * INEFFICIENT: Creating new objects/arrays in loops
 */
function inefficientObjectCreation(count) {
    const results = [];
    for (let i = 0; i < count; i++) {
        results.push({
            id: i,
            timestamp: new Date(),  // New Date object each iteration
            data: []
        });
    }
    return results;
}

/**
 * INEFFICIENT: Using try-catch in performance-critical code
 */
function inefficientErrorHandling(arr) {
    const results = [];
    for (let i = 0; i < arr.length; i++) {
        try {
            results.push(arr[i] * 2);  // Try-catch in hot loop
        } catch (e) {
            console.error(e);
        }
    }
    return results;
}

// Example usage
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        inefficientDOMManipulation,
        inefficientArrayIteration,
        inefficientFunctionCreation,
        inefficientArrayDeletion,
        inefficientMultipleIterations,
        inefficientStringConcat,
        inefficientMembershipTest,
        inefficientNestedLoops,
        inefficientVariableScope,
        inefficientSyncOperation,
        inefficientObjectCreation,
        inefficientErrorHandling
    };
}

File: examples/inefficient_example.py
"""
Example of inefficient code patterns that should be avoided.
This file demonstrates common performance issues.
"""

import time


def inefficient_list_concatenation(items):
    """
    INEFFICIENT: Using + operator in a loop for string concatenation.
    Time Complexity: O(n²) due to string immutability
    """
    result = ""
    for item in items:
        result = result + str(item) + ","
    return result


def inefficient_list_search(data, target):
    """
    INEFFICIENT: Using list for membership testing in a loop.
    Time Complexity: O(n) for each lookup
    """
    found_items = []
    for item in data:
        if item in found_items:
            continue
        found_items.append(item)
    return found_items


def inefficient_nested_loops(list1, list2):
    """
    INEFFICIENT: Nested loops without optimization.
    Time Complexity: O(n²)
    """
    matches = []
    for item1 in list1:
        for item2 in list2:
            if item1 == item2:
                matches.append(item1)
    return matches


def inefficient_file_reading(filename):
    """
    INEFFICIENT: Reading entire file into memory at once.
    Memory: O(n) where n is file size
    """
    with open(filename, 'r') as f:
        content = f.read()
        lines = content.split('\n')
        return [line.upper() for line in lines]


def inefficient_data_processing(data):
    """
    INEFFICIENT: Multiple iterations over the same data.
    """
    # First pass: filter
    filtered = []
    for item in data:
        if item > 0:
            filtered.append(item)
    
    # Second pass: square
    squared = []
    for item in filtered:
        squared.append(item ** 2)
    
    # Third pass: sum
    total = 0
    for item in squared:
        total += item
    
    return total


def inefficient_dictionary_lookup(data):
    """
    INEFFICIENT: Repeatedly checking if key exists instead of using get().
    """
    results = {}
    for item in data:
        if item in results:
            results[item] = results[item] + 1
        else:
            results[item] = 1
    return results


class InefficientDataStructure:
    """
    INEFFICIENT: Using inappropriate data structure for the use case.
    """
    def __init__(self):
        self.items = []
    
    def add_item(self, item):
        """O(n) lookup to check duplicates"""
        if item not in self.items:
            self.items.append(item)
    
    def has_item(self, item):
        """O(n) lookup"""
        return item in self.items
    
    def remove_item(self, item):
        """O(n) lookup and removal"""
        if item in self.items:
            self.items.remove(item)


def inefficient_global_variable():
    """
    INEFFICIENT: Using global variables that can cause cache misses
    and make code harder to optimize.
    """
    global global_counter
    global_counter = 0
    for i in range(1000000):
        global_counter += 1
    return global_counter


# Global variable (avoid this pattern)
global_counter = 0


if __name__ == "__main__":
    # Example usage
    print("Demonstrating inefficient patterns...")
    
    # Test string concatenation
    start = time.time()
    result = inefficient_list_concatenation(range(1000))
    print(f"String concatenation: {time.time() - start:.4f}s")
    
    # Test list search
    start = time.time()
    data = list(range(100)) * 10
    result = inefficient_list_search(data, 50)
    print(f"List search: {time.time() - start:.4f}s")
